coordinator/src/notification_service.rs

```diff
@@ -130,3 +144,205 @@ async fn send_notification<'a>(
+            &notification_sender,
+        )
+        .await;
+
+        let received_notifications = receive_all_notifications(&mut notification_receiver);
+
+        assert_eq!(received_notifications.len(), 0);
+    }
+
+    #[tokio::test]
+    async fn test_deliving_notifications_before_expiry() {
+        init_tracing_for_test();
+
+        let position_far_from_expiring = far_from_expiry_position();
+        let position_soon_to_expire = soon_expiring_position();
+        let ancient_position = ancient_expired_position(); // too old to send notification
+        let (notification_sender, mut notification_receiver) = mpsc::channel(100);
+
+        send_expiry_notifications_if_applicable(
+            &[
+                position_far_from_expiring,
+                position_soon_to_expire.clone(),
+                ancient_position,
+            ],
+            &notification_sender,
+        )
+        .await;
+
+        let received_notifications = receive_all_notifications(&mut notification_receiver);
+
+        assert_eq!(received_notifications.len(), 1);
+
+        let notification = received_notifications.first().unwrap();
+        assert_eq!(
+            notification.notification_kind,
+            NotificationKind::PositionSoonToExpire
+        );
+        assert_eq!(notification.user_fcm_token, position_soon_to_expire.1);
+    }
+
+    #[tokio::test]
+    async fn send_only_recently_expired_notifications() {
+        init_tracing_for_test();
+
+        let position_far_from_expiring = far_from_expiry_position();
+        let position_just_expired = just_expired_position();
+        let ancient_position = ancient_expired_position(); // too old to send notification
+        let (notification_sender, mut notification_receiver) = mpsc::channel(100);
+
+        send_expiry_notifications_if_applicable(
+            &[
+                position_far_from_expiring,
+                position_just_expired.clone(),
+                ancient_position,
+            ],
+            &notification_sender,
+        )
+        .await;
+
+        let received_notifications = receive_all_notifications(&mut notification_receiver);
+
+        assert_eq!(received_notifications.len(), 1);
+
+        let notification = received_notifications.first().unwrap();
+        assert_eq!(
+            notification.notification_kind,
+            NotificationKind::PositionExpired
+        );
+        assert_eq!(notification.user_fcm_token, position_just_expired.1);
+    }
+}
+
+/// Load all recent positions with the DB with associated fcm tokens and send push
+/// notification about expiring/expired positions if needed.
+pub async fn query_and_send_position_notifications(
+    conn: &mut PgConnection,
+    notification_sender: &mpsc::Sender<Notification>,
+) -> Result<()> {
+    let users = user::all(conn)?;
+
+    let positions_with_fcm_tokens = positions::Position::get_all_positions_with_expiry_after(
```

Comment:
🔧 I this could be nicely done with an inner join!

At least we should only load users which have an expired position.

Reply:
I mentioned that in the comment, but I don't think the optimisation is worthwhile at this stage?
🤔 especially if we're doing it very infrequently? (potentially, if we rework according to your other suggestions, we would only do this during the rollover weekend)